api/campaigns/class-maybe-show-campaign.php

```diff
@@ -118,10 +133,58 @@ function ( $post_data ) use ( $visit ) {
 
 			$response[ $campaign->id ] = $campaign_should_be_shown;
 		}
+
 		$this->response = $response;
 		$this->respond();
 	}
 
+	/**
+	 * Get the highest-priority segment that matches the client.
+	 *
+	 * @param object      $all_segments All segments.
+	 * @param string      $client_id Client ID.
+	 * @param string      $referer_url URL of the page performing the API request.
+	 * @param string      $page_referer_url URL of the referrer of the frontend page that is making the API request.
+	 * @param object|bool $view_as_spec View as spec.
+	 *
+	 * @return string|null ID of the best matching segment, or null if the client matches no segment.
+	 */
+	public function get_best_priority_segment( $all_segments = [], $client_id, $referer_url = '', $page_referer_url = '', $view_as_spec = false ) {
+		$client_data           = $this->get_client_data( $client_id );
+		$best_segment_priority = PHP_INT_MAX;
+		$best_priority_segment = null;
+
+		foreach ( $all_segments as $segment_id => $segment ) {
+			// If using "view as" feature, automatically make that the matching segment. Otherwise, find the matching segment with the highest priority.
+			if ( $view_as_spec && isset( $view_as_spec['segment'] ) ) {
+				if ( $view_as_spec['segment'] === $segment_id ) {
+					$best_priority_segment = $segment_id;
+				}
+			} else {
+				// Determine whether the client matches the segment criteria.
+				$segment                = Campaign_Data_Utils::canonize_segment( $segment );
+				$client_matches_segment = Campaign_Data_Utils::does_client_match_segment(
+					$segment,
+					$client_data,
+					$referer_url,
+					$page_referer_url
+				);
+
+				// Find the matching segment with the highest priority.
+				if ( $client_matches_segment ) {
+					$segment_priority = isset( $segment->priority ) ? $segment->priority : $best_segment_priority;
```

Comment:
When would a segment not have a priority, and why is the `$best_priority_segment` value the default in this case? If it's about a default value, perhaps the `Campaign_Data_Utils::canonize_segment` would be a more suitable place for that.

Reply:
To answer your question, I just set it to `$best_segment_priority` so that when the comparison happens in the next line, a segment with no priority is never possibly treated as having a better priority than another segment.

But I think you're right that it's unnecessary, because a segment should never not have a priority (and we ensure this in the segmentation class). Removed the check and default value in 1b88cf7.